Framework/Algorithms/src/ApplyInstrumentToPeaks.cpp

```diff
@@ -0,0 +1,67 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidAlgorithms/ApplyInstrumentToPeaks.h"
+#include "MantidDataObjects/PeaksWorkspace.h"
+
+namespace Mantid {
+namespace Algorithms {
+
+DECLARE_ALGORITHM(ApplyInstrumentToPeaks)
+
+using namespace Mantid::Kernel;
+using namespace Mantid::API;
+using namespace Mantid::DataObjects;
+
+void ApplyInstrumentToPeaks::init() {
+  declareProperty(std::make_unique<WorkspaceProperty<PeaksWorkspace>>("InputWorkspace", "", Direction::Input),
+                  "Input peaks workspace.");
+  declareProperty(std::make_unique<WorkspaceProperty<Workspace>>("InstrumentWorkspace", "", Direction::Input,
+                                                                 PropertyMode::Optional),
+                  "Workspace from which the instrument will be copied from. If none is provided then the instrument on "
+                  "the input workspace is used.");
+  declareProperty(std::make_unique<WorkspaceProperty<PeaksWorkspace>>("OutputWorkspace", "", Direction::Output),
+                  "Output peaks workspace.");
+}
+
+void ApplyInstrumentToPeaks::exec() {
+  /// Peak workspace to integrate
+  PeaksWorkspace_sptr inputWS = getProperty("InputWorkspace");
+
+  /// Output peaks workspace, create if needed
+  PeaksWorkspace_sptr outputWS = getProperty("OutputWorkspace");
+  if (outputWS != inputWS)
+    outputWS = inputWS->clone();
+
+  Mantid::Geometry::Instrument_const_sptr instrument;
+
+  Workspace_sptr instWS = this->getProperty("InstrumentWorkspace");
+  if (instWS) {
+    ExperimentInfo_sptr ei = std::dynamic_pointer_cast<ExperimentInfo>(instWS);
+    if (!ei)
+      throw std::invalid_argument("Wrong type of workspace");
+    instrument = ei->getInstrument();
+    outputWS->setInstrument(instrument);
+  } else {
+    instrument = outputWS->getInstrument();
+  }
+
+  for (int i = 0; i < outputWS->getNumberPeaks(); ++i) {
+    Peak &peak = outputWS->getPeak(i);
+    const auto tof = peak.getTOF();
+    peak.setInstrument(instrument);
+    peak.setDetectorID(peak.getDetectorID());
+    const double energy = pow((peak.getL1() + peak.getL2()) / tof * 1e6, 2) * PhysicalConstants::NeutronMass / 2.0 /
+                          PhysicalConstants::meV;
+    peak.setInitialEnergy(energy);
+    peak.setFinalEnergy(energy);
```

Comment:
Just out of curiosity, is this part of the code equivalent to the following?
```c++
    Units::Wavelength wl;
    wl.initialize(pk.getL1(), 0,
                  {{UnitParams::l2, pk.getL2()},
                   {UnitParams::twoTheta, pk.getScattering()},
                   {UnitParams::efixed, pk.getInitialEnergy()}});
    pk.setWavelength(wl.singleFromTOF(tof + dT0));
```

Reply:
Should be except there is no `dT0` factor. I think I will change my calculation to use Mantid Units, but still just to energy since peak just transforms it to energy anyway.